### ISSUE TITLE
Improve how Wasp handles and reports unavailable ports

### ISSUE THREAD

**Issue by sodic** [wasp-lang/wasp#2427]
Version: 0.15.2

Steps to reproduce:
- Start a process (not Wasp) that listens on port 3001. For example: `python -m http.server 3000`
- Execute `wasp start`.
- Wasp will fail and tell you _"Port 3001 is taken"_ (not 3000).

What happens:
- The client sees port 3000 is taken and automatically falls back to 3001.
- The server sees port 3001 is taken and crashes with the incorrect error.

It's a confusing experience because the client's message that says _"Port 3000 is in use, trying another one..."_ is printed earlier and gets lost in the noise, while the server's message that says _"Port 3001 is already in use"_ is the last (and possibly the only) thing the user sees.

@Martinsos Don't we already have an issue for this? I couldn't find it.

**Comment by infomiho** [comment]
This could be a good option to get more predictable behavior: https://vite.dev/config/server-options#server-strictport

**Comment by kvsm** [comment]
> What happens:
> 
> * The client sees port 3000 is taken and automatically falls back to 3001.
> * The server sees port 3001 is taken and crashes with the incorrect error.

What's the 'correct' behaviour here? Should the server fall back to 3002, or should it fail with a correct error?

**Comment by Martinsos** [comment]
> > What happens:
> > 
> > * The client sees port 3000 is taken and automatically falls back to 3001.
> > * The server sees port 3001 is taken and crashes with the incorrect error.
> 
> What's the 'correct' behaviour here? Should the server fall back to 3002, or should it fail with a correct error?

Good question!

Related and relevant issue: https://github.com/wasp-lang/wasp/issues/730 .
Also somewhat related (but less) issues: https://github.com/wasp-lang/wasp/issues/1858 and https://github.com/wasp-lang/wasp/issues/1049 .

So question is, what is the bhevaiour we want? Either we want stuff to fail in an obvious manner, so user knows what happened and can free the ports, or we can make it so that ports are changed till they pick up a free one, so chosen automatically. Which one of these two, I am not sure. I guess the second one sounds better, but first one is easier and more predictable. Maybe both -> allow users some control over it, maybe they can specify the ports themselves, and if not, we pick them automatically.

What do other frameworks/solutions do?

**Comment by kvsm** [comment]
> What do other frameworks/solutions do?

I checked:
`Next`, which falls over to the next port with a warning message.
`serve`, which falls over to a random port in the 63k range with a warning message.
`Astro`, which appears to take the port whether it is in use or not (??)

I agree that falling over in as obvious a way as possible is probably the best DX, but for the meantime I'll implement @infomiho 's suggestion.

**Comment by sodic** [comment]
> Astro, which appears to take the port whether it is in use or not (??)

Yes! I noticed this happening with other tools as well (e.g. Docusaurus). Everything I know about computers tells me that shouldn't be possible. I have no idea what's going on 😄 

> I agree that falling over in as obvious a way as possible is probably the best DX, but for the meantime I'll implement Miho's suggestion

Excellent, thanks!

**Comment by cprecioso** [comment]
From https://github.com/wasp-lang/wasp/pull/2796#discussion_r2158645082

We can use `--strictPort` (from Vite docs: exit if specified port is already in use)

**Comment by sodic** [comment]
Closed with https://github.com/wasp-lang/wasp/pull/4585 and https://github.com/wasp-lang/wasp/pull/4586
